stan/math/gpu/cholesky_decompose.hpp

```diff
@@ -0,0 +1,127 @@
+inline matrix_gpu cholesky_decompose_recursion(matrix_gpu& A,
+                                               const int block = 100,
+                                               const int divider = 2,
+                                               const int min_block = 100) {
+  matrix_gpu L(A.rows(), A.cols());
+  if (block <= min_block || divider <= 1) {
+    try {
+      opencl_kernels::cholesky_decompose(cl::NDRange(A.rows()),
+                                         cl::NDRange(A.rows()), A.buffer(),
+                                         L.buffer(), A.rows());
+    } catch (const cl::Error& e) {
+      check_opencl_error("cholesky_decompose", e);
+    }
+  } else {
+    L = stan::math::cholesky_decompose(A, block / divider);
+  }
+  return L;
+}
+}  // namespace internal
+/**
+ * Return the lower-triangular Cholesky factor (i.e., matrix
+ * square root) of the specified square, symmetric matrix.
+ * The return value \f$L\f$ will be a lower-traingular matrix such that the
+ * original matrix \f$A\f$ is given by
+ * <p>\f$A = L \times L^T\f$.
+ * The Cholesky decomposition is computed on the GPU. This algorithm is
+ * recursive, where The parameters <code>block</code>, <code>divider</code>, and
+ *  <code>min_block</code> act as tuning parameters for the recursive step of
+ *  the GPU based Cholesky decompostion. The matrix is subset by the
+ *  <code>block</code> size, and if the <code>block</code> size is less than
+ * <code>min_block</code> then the cholesky decomposition on the GPU is computed
+ * using that submatrix. If <code>block</code> is greater than
+ * <code>block_size</code> then <code>cholesky_decompose</code> is run again
+ * with <code>block</code> equal to <code>block/divider</code>. Once the
+ * Cholesky Decomposition is computed, the full matrix cholesky is created
+ * by propogating the cholesky forward as given in the reference report below.
+ *
+ * For a full guide to how this works
+ * see the Cholesy decompostion chapter in the  reference report
+ * <a href="https://goo.gl/6kWkJ5"> here</a>.
+ * @param A Symmetric matrix on the GPU.
+ * @param block Size of the block used to compute the cholesky decomposition.
+ * @param divider Proportion to divide the submatrix by at each recursive step.
+ * @param min_block The amount that block is checked against to decide
+ *  whether to continue the recursion or to perform the cholesky.
+ * @return Square root of matrix on the GPU.
+ * @throw std::domain_error if m is not
+ *  positive definite (if m has more than 0 elements)
+ */
+inline matrix_gpu cholesky_decompose(matrix_gpu& A, const int block,
+                                     const int divider, const int min_block) {
+  auto offset = 0;
+  // NOTE: The code in this section follows the naming conventions
+  // in the report linked in the docs.
+  matrix_gpu A_11(block, block);
+  // Repeats the blocked cholesky decomposition until the size of the remaining
+  // submatrix is smaller or equal to the block size
```

Comment:
I think the comment is accurate, but worded weirdly. I'll clean this up

Reply:
Actually reading that comment it's better to change the comment on line 108 so that it references the remaining submatrix. Then readers can get that the part here runs until the block size exceeds the size of the matrix then the part of the matrix remaining is computed below